### PR TITLE
Add documentation about client-side load balancing and service discovery

### PR DESCRIPTION
**Pull request by trustin**
Modifications:
- Add client-service-discovery.rst which explains client-side load
  balancing and service discovery
- Update `overrides.css` so that the menu items look better, especially
  monospaced ones

### CODE REVIEW

**Comment by trustin** [comment]
Screenshot: https://screenshots.firefox.com/q3eChm85RtnM34ug/null

File: site/src/sphinx/client-service-discovery.rst [hunk 1]
```diff
@@ -0,0 +1,237 @@
+.. _client-service-discovery:
+
+Client-side load balancing and service discovery
+================================================
+
+You can configure an Armeria client to distribute its requests to more than one server autonomously, unlike
+traditional server-side load balancing where the requests goes through a dedicated load balancer such as
```
Comment (imasahiro) [inline site/src/sphinx/client-service-discovery.rst]: `the requests goes through` -> `the requests go through`

File: site/src/sphinx/client-service-discovery.rst [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    EndpointGroupRegistry o-right- "*" EndpointGroup
+    EndpointGroup o-right- "*" Endpoint
+    @enduml
+
+.. _creating-endpoint-group:
+
+
+Creating an ``EndpointGroup``
+-----------------------------
+There are various :api:`EndpointGroup` implementations provided out of the box, but let's start simple with
```
Comment (minwoox) [inline site/src/sphinx/client-service-discovery.rst]: let's simply start with?

File: site/src/sphinx/client-service-discovery.rst [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+Registering an ``EndointGroup``
+-------------------------------
+An :api:`EndpointGroup` becomes visible by a client such as :api:`HttpClient` only after it's registered in
+:api:`EndpointGroupRegistry`. You need to specify 2 more elements to register an :api:`EndpointGroup`:
+
+- The name of the :api:`EndpointGroup`
+- An :api:`EndpointSelectionStrategy` that determines which :api:`Endpoint` is used for each request
+
+  - Use ``EndpointSelectionStrategy.WEIGHTED_ROUND_ROBIN`` for weighted round robin.
+  - Use ``EndpointSelectionStrategy.ROUND_ROBIN`` for unweighted rounb robin.
```
Comment (minwoox) [inline site/src/sphinx/client-service-discovery.rst]: rounb?

File: site/src/sphinx/client-service-discovery.rst [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+Once an :api:`EndpointGroup` is registered, you can use its name in the authority part of a URI:
+
+.. code-block:: java
+
+    // Create an HTTP client that sends requests to the 'search_engines' group.
+    HttpClient client = HttpClient.of("https://group:search_engines/");
+
+    // Send a GET request to each search engine.
+    List<CompletableFuture<? >> futures = new ArrayList<>();
```
Comment (minwoox) [inline site/src/sphinx/client-service-discovery.rst]: `<? >>` is intentional(the space after `?`)?

File: site/src/sphinx/client-service-discovery.rst [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    is called *dynamic endpoint group*.
+
+
+Removing unhealthy ``Endpoint`` with ``HttpHealthCheckedEndpointGroup``
+-----------------------------------------------------------------------
+:api:`HttpHealthCheckedEndpointGroup` decorates an existing :api:`EndpointGroup` to filter out the unhealthy
+:apiplural:`Endpoint` from it so that a client has less chance of sending its requests to the unhealthy
+:apiplural:`Endpoint`. It determines the healthiness by sending so called 'health check request' to each
+:api:`Endpoint`, which is by default a simple ``GET`` request to a certain path. If an :api:`Endpoint`
+responds with non-200 status code or does not respond in time, it will be marked unhealthy and thus
```
Comment (minwoox) [inline site/src/sphinx/client-service-discovery.rst]: marked as unhealthy?

File: site/src/sphinx/client-service-discovery.rst [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    StaticEndpointGroup group = new StaticEndpointGroup(
+        Endpoint.of("192.168.0.1", 80),
+        Endpoint.of("192.168.0.2", 80));
+
+    // Decorate the EndpointGroup with HttpHealthCheckedEndpointGroup
+    // that sends HTTP health check requests to '/internal/l7check' every 10 seconds.
+    HttpHealthCheckedEndpointGroup healthCheckedGroup =
+            new HttpHealthCheckedEndpointGroupBuilder(searchEngineGroup, "/internal/l7check")
+                    .protocol(SessionProtocol.HTTP)
+                    .retryInternal(Duration.ofSeconds(10))
```
Comment (minwoox) [inline site/src/sphinx/client-service-discovery.rst]: retryInterval?

File: site/src/sphinx/client-service-discovery.rst [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+They refresh the :api:`Endpoint` list automatically, respecting TTL values, and retry when DNS queries fail.
+
+:api:`DnsAddressEndpointGroup` is useful when accessing an external service with multiple public IP addresses:
+
+.. code-block:: java
+
+    DnsAddressEndpointGroup group =
+            new DnsAddressEndpointGroupBuilder("www.google.com")
+                    // Do not refresh more often than every 10 seconds and
```
Comment (minwoox) [inline site/src/sphinx/client-service-discovery.rst]: Do not refresh -> Refresh ??

File: site/src/sphinx/client-service-discovery.rst [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+`L4 and L7 switches <https://en.wikipedia.org/wiki/Multilayer_switch#Layer_4%E2%80%937_switch,_web_switch,_or_content_switch>`_.
+
+There are 4 elements involved in client-side load balancing in Armeria:
+
+- :api:`Endpoint` represents an individual host (with an optional port number) and its weight.
+- :api:`EndpointGroup` represents a set of :apiplural:`Endpoint`.
+- :api:`EndpointGroupRegistry` is a global registry of :apiplural:`EndpointGroup` where each
+  :api:`EndpointGroup` is identified by its unique name.
+- A user specifies the target group name in a URI,
+  e.g. ``http://group:my_group/`` where ``my_group`` is the group name.
```
Comment (huydx) [inline site/src/sphinx/client-service-discovery.rst]: how about 

```
e.g.  ``http://group:group_name``  (examples: ``http://group:google/`` or ``http://group:facebook/``)
```

File: site/src/sphinx/client-service-discovery.rst (hunk 2 above)
Comment (trustin) [reply site/src/sphinx/client-service-discovery.rst]: I guess 'start simple' is OK here. e.g. think big, act small

File: site/src/sphinx/client-service-discovery.rst (hunk 4 above)
Comment (trustin) [reply site/src/sphinx/client-service-discovery.rst]: No. Was a mistake :sweat_smile:

File: site/src/sphinx/client-service-discovery.rst (hunk 7 above)
Comment (trustin) [reply site/src/sphinx/client-service-discovery.rst]: Oops! Thanks

File: site/src/sphinx/client-service-discovery.rst (hunk 8 above)
Comment (trustin) [reply site/src/sphinx/client-service-discovery.rst]: `e.g.` means 'for example', so it seems weird to have an example in an example.

Comment (trustin) [reply site/src/sphinx/client-service-discovery.rst]: Added a little bit more information for clarity

File: site/src/sphinx/client-service-discovery.rst [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+    // Create an EndpointGroup with 2 Endpoints.
+    StaticEndpointGroup group = new StaticEndpointGroup(
+        Endpoint.of("192.168.0.1", 80),
+        Endpoint.of("192.168.0.2", 80));
+
+    // Decorate the EndpointGroup with HttpHealthCheckedEndpointGroup
+    // that sends HTTP health check requests to '/internal/l7check' every 10 seconds.
+    HttpHealthCheckedEndpointGroup healthCheckedGroup =
+            new HttpHealthCheckedEndpointGroupBuilder(searchEngineGroup, "/internal/l7check")
```
Comment (huydx) [inline site/src/sphinx/client-service-discovery.rst]: should `searchEngineGroup` be `group`?

**Comment by huydx** [comment]
Read the docs raise a question for me, what happen if we mixed up 2 dynamic EndpointGroup together. Ex: if we used both Dns group and HttpHealthChecked group, is it still work as expected when 2 group try to modify the list at the same time?

**Comment by trustin** [comment]
@huydx Yes, you're right. Let me add a little bit more make it more explicit.

File: site/src/sphinx/client-service-discovery.rst (hunk 9 above)
Comment (trustin) [reply site/src/sphinx/client-service-discovery.rst]: Oh, yes. Thanks!

**Comment by huydx** [comment]
LGTM 💯

**Comment by trustin** [comment]
Thanks, reviewers!

**Pull request merged**